src/mbgl/util/http_timeout.cpp

```diff
@@ -0,0 +1,39 @@
+#include <mbgl/util/http_timeout.hpp>
+
+namespace mbgl {
+namespace http {
+
+Duration errorRetryTimeout(Response::Error::Reason failedRequestReason, uint32_t failedRequests, optional<Timestamp> retryAfter) {
+
+    if (failedRequestReason == Response::Error::Reason::Server) {
+        // Retry after one second three times, then start exponential backoff.
+        return Seconds(failedRequests <= 3 ? 1 : 1u << std::min(failedRequests - 3, 31u));
+    } else if (failedRequestReason == Response::Error::Reason::Connection) {
+        // Immediate exponential backoff.
+        assert(failedRequests > 0);
+        return Seconds(1u << std::min(failedRequests - 1, 31u));
+    } else if (failedRequestReason == Response::Error::Reason::RateLimit) {
+        if (retryAfter) {
+            return *retryAfter - util::now();
+        } else {
+            //Default
+            return Seconds(5);
```

Comment:
Move the magic constant to `./include/mbgl/util/constants.hpp`. We already have similar things like `CLOCK_SKEW_RETRY_TIMEOUT` there.